multicore_runtime/wsgi_test.py

```diff
@@ -0,0 +1,284 @@
+    appinfo.URLMap(url='/env', script='wsgi_test.dump_os_environ'),
+    appinfo.URLMap(url='/sortenv', script='wsgi_test.sort_os_environ_keys'),
+    appinfo.URLMap(url='/setenv', script='wsgi_test.add_to_os_environ'),
+    appinfo.URLMap(url='/wait', script='wsgi_test.wait_on_global_event'),
+    appinfo.URLMap(url='/login', script='wsgi_test.hello_world',
+                   login=appinfo.LOGIN_REQUIRED),
+    appinfo.URLMap(url='/admin', script='wsgi_test.hello_world',
+                   login=appinfo.LOGIN_ADMIN),
+    appinfo.URLMap(url='/favicon.ico',
+                   static_files='test_statics/favicon.ico',
+                   upload='test_statics/favicon.ico'),
+    appinfo.URLMap(url='/faketype.ico', static_files='test_statics/favicon.ico',
+                   mime_type='application/fake_type',
+                   upload='test_statics/favicon.ico'),
+    appinfo.URLMap(url='/wildcard_statics/(.*)',
+                   static_files=r'test_statics/\1',
+                   upload='test_statics/(.*)'),
+    appinfo.URLMap(url='/static_dir',
+                   static_dir='test_statics'),
+    ]
+HELLO_STRING = 'Hello World!'
+
+FAKE_ENV_KEY = 'KEY'
+FAKE_ENV_VALUE = 'VALUE'
+FAKE_USER_EMAIL = 'test@example.com'
+BAD_USER_EMAIL = 'bad@example.com'
+
+FAKE_APPINFO_EXTERNAL = MagicMock(handlers=FAKE_HANDLERS,
+                                  env_variables={FAKE_ENV_KEY: FAKE_ENV_VALUE,
+                                                 'USER_EMAIL': BAD_USER_EMAIL})
+
+FAKE_APPENGINE_CONFIG = MagicMock(
+    server_software='server', partition='partition', appid='appid',
+    module='module', instance='instance', major_version='major',
+    minor_version='minor', default_ticket='ticket')
+
+
+# Global event flags used for concurrency tests.
+concurrent_request_is_started = threading.Event()
+concurrent_request_should_proceed = threading.Event()
+
+# These tests will deliberately cause ERROR level logs, so let's disable them.
+logging.basicConfig(level=logging.CRITICAL)
+
+
+@Request.application
+def hello_world(request):  # pylint: disable=unused-argument
+  return Response(HELLO_STRING)
+
+
+@Request.application
+def dump_os_environ(request):  # pylint: disable=unused-argument
+  return Response(json.dumps(dict(os.environ)))
+
+
+@Request.application
+def add_to_os_environ(request):  # pylint: disable=unused-argument
+  os.environ['ENVIRONMENT_MODIFIED'] = 'TRUE'
+  return Response(json.dumps(dict(os.environ)))
+
+
+@Request.application
+def wait_on_global_event(request):  # pylint: disable=unused-argument
+  concurrent_request_is_started.set()
+  concurrent_request_should_proceed.wait()
+  return Response(json.dumps(dict(os.environ)))
+
+
+@Request.application
+def sort_os_environ_keys(request):  # pylint: disable=unused-argument
+  # See test_env_sort method for explanation.
+  resp = ''
+  for name in sorted(os.environ):
+    resp += '%s=%s\n' % (name, os.environ[name])
```

Comment:
Concatenation of strings is usually poor... Instead, use `'\n'.join(...)` or `''.join(...)` with either a list you populate or a generator.

Note that when you use an operation such as `join` it takes a generator so you can skip the `[` and `]` to turn it into a list:

```python
  return Response(''.join('%s=%s\n' % (name, value) for name, value in sorted(os.environ.iteritems()))
```

(You can also get both name and value from `os.environ` at once with `iteritems()`. And `sorted()` should be happy with it, it will sort a tuple by the first element and second only if the first is the same. Using `sorted()` on the strings would probably yield a very similar, probably exactly the same result, in this particular case.)